spring-integration-file/src/main/java/org/springframework/integration/file/splitter/FileSplitter.java

```diff
@@ -224,6 +226,23 @@ public Object next() {
 		}
 	}
 
+	@Override
+	protected void addHeaders(Message<?> message, AbstractIntegrationMessageBuilder<?> builder) {
+		if (message.getPayload() instanceof File) {
+			File file = (File) message.getPayload();
+			builder
+				.setHeaderIfAbsent(FileHeaders.ORIGINAL_FILE, file)
+				.setHeaderIfAbsent(FileHeaders.FILENAME, file.getName());
+
+		}
+		else if (message.getPayload() instanceof String) {
+			File file = new File((String) message.getPayload());
```

Comment:
Another my concern here that you expose `AbstractIntegrationMessageBuilder` to end-users and it is dangerous: no guaranty what he is going to do with that.
We had with you similar discussion some time ago around `TcpMessageMapper.addCustomHeaders`.